src/target/riscv/riscv-013_reg.c

```diff
@@ -57,17 +56,6 @@ static int riscv013_reg_set(struct reg *reg, uint8_t *buf)
 			buf_get_u64(buf, 0, reg->size) == 0)
 		return ERROR_OK;
 
-	if (reg->number == GDB_REGNO_TDATA1 ||
-			reg->number == GDB_REGNO_TDATA2) {
-		r->manual_hwbp_set = true;
-		/* When enumerating triggers, we clear any triggers with DMODE set,
-		 * assuming they were left over from a previous debug session. So make
-		 * sure that is done before a user might be setting their own triggers.
-		 */
-		if (riscv_enumerate_triggers(target) != ERROR_OK)
-			return ERROR_FAIL;
-	}
-
```

Comment:
I think we should reserve the trigger (together with an INFO or a WARNING print if it isn't already reserved) if the user touches it via `reg_set()`

Trough, this raises a question if a watchpoint is set on said trigger, if we should remove it from memory.

What do others think?

Reply:
I'd like to avoid it.
1. This will require reading `tselect` to know which trigger to reserve. IMHO `reg` command should strive to minimize the number of operations it does.
2. TBH, I think Trigger Module registers should probably be marked as hidden and should not be accessible to user  by default.

Reply:
> What do others think?

1. My vote is to always print big ugly error message when user tries to access trigger module registers explicitly. 
2. I would go as far as to render trigger module registers as unavailable to the user by default.

Otherwise it may create way too much hassle because of some obscure corner case when a user wants to mess up with the trigger module while having bp/wp set. I don't think any sane person want to do that. If user is crazy enough to mess with this stuff - he can unlock them an do whatever she/he wants (without any baby-sitting from OpenOCD side).

That being said I would recommend to file an issue for this functionality and merge this patch as is, since in my opinion this patch makes things a little bit better compared to what we have now.

Reply:
I am happy to see the `manual_hwbp_set` heuristics removed, since it silently changed the behavior of enabling/disabling triggers when user touched any of the trigger registers.

> I think we should reserve the trigger if (...) What do others think?

My preference would be to keep the trigger CSRs normally visible to the user, and don't perform any extra action if the user modifies any of the trigger CSRs. OpenOCD is a low-level debug tool, and as such it gives the user generally a lot of freedom about what it can do, and at the same time it requires cautious use, otherwise things may go awry.

I'd apply the same logic to the trigger manipulation, and not try to make OpenOCD any more "clever" - just do what the user says. :)